README.md
# lita-pulp

A lita handler plugin for pulp server operations.

## Installation

Add lita-pulp to your Lita instance's Gemfile:

``` ruby
gem "lita-pulp"
```

## Configuration

Configuration options

```ruby
Lita.configure do |config|
  config.handlers.pulp.url="https://pulp.co.epi.web"
  config.handlers.pulp.api_path="/pulp/api/v2/"
  config.handlers.pulp.username="admin"
  config.handlers.pulp.password="admin"
  config.handlers.pulp.verify_ssl=false #optional default to false
end
```

## Usage

It can do following:
- List all rpm repositories
- List all puppet repositories
- Show a repository details
- Search rpm package
- Search puppet module
- Copy rpm package from one repository to another
- Copy puppet module from one repository to another
- Create rpm/puppet repository
- Delete a repository


CHANGELOG.md
# Lita-Pulp Change Log

## 0.1.4
* Remove requirement of pulp_version configuration parameter, use dynamic reflection
* Add configuration for pulp_admins_group, default to :pulp_admins, repository delete/create permitted only to these groups

## 0.1.3
* create puppet rpm repository
* delete a repository


spec/spec_helper.rb
require 'lita-keyword-arguments'
require "lita-pulp"
require "lita/rspec"
require "docker/compose"

# A compatibility mode is provided for older plugins upgrading from Lita 3. Since this plugin
# was generated with Lita 4, the compatibility mode should be left disabled.
Lita.version_3_compatibility_mode = false

require 'vcr'


VCR.configure do |c|
  c.default_cassette_options = {
    :match_requests_on => [:method, :uri, :host, :headers, :query, :body, :path]
  }
  c.ignore_request do |request |
    uri = request.uri
    ignore = false
    method = request.method
    if uri.end_with?('repositories/') && method.to_s == 'post'
      ignore = true
    end
    ignore == true
  end
  c.cassette_library_dir = 'spec/fixtures/cassettes'
  c.hook_into :webmock
  c.configure_rspec_metadata!
end


RSpec.configure do |config|
  config.before :suite do
    puts "before suite, set up"
    work_dir = File.expand_path('../fixtures', __FILE__)
    if ENV['LITA_provision'] != 'no'
      @compose = Docker::Compose::Session.new dir:work_dir
      @compose.up detached:true, no_build: true
    end
  end
  config.after :suite do
    puts "after suite, clean up"
    if ENV['LITA_destroy'] != 'no'
      @compose.stop
      sleep 2
      @compose.rm force:true, volumes:true
    end
  end
  config.before do
    registry.register_handler(Lita::Handlers::Pulp)
    registry.register_hook(:trigger_route, Lita::Extensions::KeywordArguments)
  end
end


spec/lita/handlers/pulp_spec.rb
require "spec_helper"

describe Lita::Handlers::Pulp, lita_handler: true do

  before do
    registry.config.handlers.pulp.url="https://pulp.co.epi.web"
    registry.config.handlers.pulp.api_path="/pulp/api/v2/"
    registry.config.handlers.pulp.username="admin"
    registry.config.handlers.pulp.password="admin"
    registry.config.handlers.pulp.verify_ssl=false
  end

  it do
    is_expected.to route_command('pulp rpm repos').to(:rpm_repos)
    is_expected.to route_command('pulp puppet repos').to(:puppet_repos)
  end

  describe '#rpm_repos' do
    it 'list rpm repos' do
      send_command("pulp rpm repos")
      puts "*********************"
      puts replies
    end
  end

  describe '#puppet_repos' do
    it 'list puppet repos' do
      send_command("pulp puppet repos")
      puts "*********************"
      puts replies
    end
  end
end


spec/lita/handlers/pulp_repo_spec.rb
require "spec_helper"

vcr_options = { :record => :new_episodes }

describe Lita::Handlers::Pulp, lita_handler: true, :vcr => vcr_options do

  before do
    registry.config.handlers.pulp.url="https://localhost:8843"
    registry.config.handlers.pulp.api_path="/pulp/api/v2/"
    registry.config.handlers.pulp.username="admin"
    registry.config.handlers.pulp.password="admin"
    registry.config.handlers.pulp.verify_ssl=false
  end

  it do
    is_expected.to route_command('pulp rpm repos').to(:rpm_repos)
    is_expected.to route_command('pulp puppet repos').to(:puppet_repos)
    is_expected.to route_command('pulp show repo test').to(:show_repo)
    is_expected.to route_command('pulp sync test').to(:repo_sync)
    is_expected.to route_command('pulp publish test').to(:repo_publish)
    is_expected.to route_command('pulp create rpm repo --repo_id rpm_repo_1 --name "test rpm repo"').to(:cmd_create_rpm_repo)
    is_expected.to route_command('pulp create puppet repo --repo_id puppet_repo_1 --name "test puppet repo"').to(:cmd_create_puppet_repo)
    is_expected.to route_command('pulp delete repo test').to(:cmd_delete_repository)
  end


  describe '#create rpm repo' do
    it 'should create repo_rpm_1 successfully' do
      send_command('pulp create rpm repo --repo_id repo_rpm_1 --name "test repo" --description "Test REPO" --relative_url "rpm/test/x86_64" ')
      puts replies
    end
    it 'should create repo_rpm_2 successfully' do
      send_command('pulp create rpm repo --repo_id repo_rpm_2 --name "test repo 2" --description "Test REPO2" --relative_url "rpm/test2/x86_64" ')
      puts replies
    end
  end

  describe '#create puppet repo' do
    it 'should create repo_puppet_1 successfully' do
      send_command('pulp create puppet repo --repo_id repo_puppet_1 --name "test puppet repo 1" --description "Test Puppet REPO1" --remove_missing ')
      puts replies
    end
    it 'should create repo_puppet_2 successfully' do
      send_command('pulp create puppet repo --repo_id repo_puppet_2 --name "test puppet repo 2" --description "Test Puppet REPO2" --remove_missing ')
      puts replies
    end
  end

  describe '#rpm_repos' do
    it 'list rpm repos' do
      send_command("pulp rpm repos")
      expect(replies.last).to match(/repo_rpm_1|repo_rpm_2/)
    end
  end

  describe '#puppet_repos' do
    it 'list puppet repos' do
      send_command("pulp puppet repos")
      #puts replies
      expect(replies.last).to match(/repo_puppet_1|repo_puppet_2/)
    end
  end

  describe '#show repo detail' do
    it 'show a rpm repo detail' do
      send_command('pulp show repo repo_rpm_1')
      expect(replies.last).to match(/id.*repo_rpm_1/)
      expect(replies.last).to match(/"http".*true/)
      expect(replies.last).to match(/"https".*false/)
    end
    it 'show a puppet repo detail' do
      send_command('pulp show repo repo_puppet_1')
      expect(replies.last).to match(/id.*repo_puppet_1/)
      expect(replies.last).to match(/"serve_http".*true/)
      expect(replies.last).to match(/"serve_https".*false/)
    end
  end

  describe '#sync repo' do
    it 'sync a repository' do
      send_command('pulp sync repo_rpm_1')
      #puts replies
      expect(replies.last).to match(/task_id/)
    end
  end

  describe '#publish repo' do
    it 'publish a repository' do
      send_command('pulp publish repo_rpm_1')
      expect(replies.last).to match(/task_id/)
    end
  end

  describe '#delete_repo' do
    it 'delete rpm repo' do
      send_command("pulp delete repo repo_rpm_2")
      #puts replies
      expect(replies.last).to match(/success/i)
    end
  end

end


lib/pulphelper/unit.rb
require 'json'
module PulpHelper
  module Unit
    def search_rpm(name, repo)
      criteria = {
        "fields" => ["name", "version", "release", "epoch", "checksum"],
        "filters" => {
          "name" => {
            "$regex" => name
          }
        },
        "sort" => [["epoch", "descending"], ["version", "descending"],  ["release", "descending"]]
      }
      #puts "search criteria #{criteria} in repo : #{repo}"
      response = client.resources.unit.search("rpm", criteria, {"include_repos" => true})
      code = response.code
      body = response.body
      #puts body
      results =[]
      case code
      when 200
        units = JSON.parse(body.to_json)
        units.each do |unit|
          result = {
            :name => unit["name"],
            :epoch => unit["epoch"],
            :version => unit["version"],
            :release => unit["release"],
            :checksum =>  unit["checksum"],
            :repos => unit["repository_memberships"]
          }
          results << result
        end
      else
        raise "Excpetion:response code = #{code}"
      end
      #filter by repo
      if repo
        results =results.select {|r|
          r[:repos].include? repo
        }
      end
      #puts "result:#{results}"
      results
    end # end search
    def search_puppet(author, name, repo)
      if author.nil?
        criteria = {
          "fields" => ["author","name", "version"],
          "filters" => {
            "name" => {
              "$regex" => name
            }
          },
          "sort" => [["author", "descending"],["name", "descending"],["version", "descending"]]
        }
      else
        criteria = {
          "fields" => ["author","name", "version"],
          "filters" => {
            "$and" => [
              {
                "name" => {
                  "$regex" => name
                }
              },
              {
                "author" => author
              }
            ]
          },
          "sort" => [["author", "descending"],["name", "descending"],["version", "descending"]]
        }
      end
      #puts "search criteria #{criteria}"
      response = client.resources.unit.search("puppet_module", criteria, {"include_repos" => true})
      code = response.code
      body = response.body
      #puts body
      results =[]
      case code
      when 200
        units = JSON.parse(body.to_json)
        units.each do |unit|
          result ={
            :name => unit["name"],
            :version => unit["version"],
            :author => unit["author"],
            :repos => unit["repository_memberships"]
          }
          results << result
        end
      else
        raise "Exception: search response code = #{code}"
      end
      #filter by repo
      if repo
        results =results.select {|r|
          r[:repos].include? repo
        }
      end
      #puts results
      results
    end

  end
end


lib/pulphelper/user.rb
require 'runcible'

module PulpHelper
  module User

    def create_user(username, password)
      response = client.resources.user.create(username, :name => username, :password => password)
      if response.code == 201
        "User created."
      else
        raise "Failed to create user, response code: #{response.code}"
      end
    end

    def get_user(username)
      response = client.resources.retrive(username)
      if response.code == 200
        response.body
      else
        raise "Failed to retrieve user, response code #{response.code}"
      end
    end

    def delete_user(username)
      response = client.resources.delete(username)
      if response.code == 200
        response.body
      else
        raise "Failed to delete user, response code #{response.code}"
      end
    end

    def change_password(password)

    end

    def assign_role(user, role)
      response = client.resources.add(role, user)
      if response.code == 200
        response.body
      else
        raise "Failed to assign role to user, response code #{response.code}"
      end
    end

    def unassign_role(user, role)
      response = client.resources.remove(role, user)
      if response.code == 200
        response.body
      else
        raise "Failed to unassign role from user, response code #{response.code}"
      end
    end

  end
end

lib/pulphelper/repo.rb
require 'json'
require 'runcible'

module PulpHelper
  module Repo
    REPO_TYPE_RPM="rpm-repo"
    REPO_TYPE_PUPPET="puppet-repo"

    YUM_IMPORTER_ID = "yum_importer"
    PUPPET_IMPORTER_ID = 'puppet_importer'

    YUM_DISTRIBUTOR_ID = "yum_distributor"
    PUPPET_DISTRIBUTOR_ID = "puppet_distributor"
    YUM_EXPORT_DISTRIBUTOR_ID="export_distributor"

    def list_repo(type)
      # use $in seems buggy
      criteria = {
        "filters" => {
          "notes._repo-type" => type
        }
      }
      #puts "criteria:#{criteria}"
      response=client.resources.repository.search(criteria)
      code=response.code
      body=response.body
      #puts "list result: #{body}"
      result=[]
      case code
      when 200
        repos=JSON.parse(body.to_json)
        #puts repos
        repos.each do |repo|
            repo_data={
              :id => repo["id"],
              :name => repo["display_name"],
              :description => repo["description"]
            }
            #puts repos
            result << repo_data
        end
      else
        raise "Exception: cannot list repository: response code :#{code}"
      end
      return result
    end#list_repo

    #yum repo
    # distributors
    #   distributor_type_id : yum_distributor
    #     auto_publish: true or false
    #     last_publish:
    #     config:
    #       http:
    #       https:
    #       relative_url
    # importers
    #    importer_type_id:  yum_importer
    #    last_sync:
    #    config:
    #      feed

    #puppet repo
    # distributors:
    #  distributor_type_id: "puppet_distributor"
    #    auto_publish:
    #    last_publish
    #    config :
    # importers:
    #   importer_type_id: "puppet_importer"
    def get_repo(repo_id)
      response=client.extensions.repository.retrieve_with_details(repo_id)
      code=response.code
      body=response.body
      case code
      when 200
        repo=JSON.parse(body.to_json)
        type = repo["notes"]["_repo-type"]
        #puts repos
        repo_data=nil
        case type
          when REPO_TYPE_RPM
            yum_distributor = repo["distributors"].select{ |d| d["distributor_type_id"] == 'yum_distributor'}[0]
            yum_importer = repo["distributors"].select{ |d| d["distributor_type_id"] == 'yum_importer'}[0]
            distributor = nil
            if  yum_distributor
              distributor = {
                :auto_publish => yum_distributor["auto_publish"],
                :last_publish => yum_distributor["last_publish"],
                :config => yum_distributor["config"]
              }
            end
            importer = nil
            if yum_importer
              importer = {
                :last_sync => yum_importer["last_sync"],
                :config => yum_importer["config"]
              }
            end

            repo_data={
              :id => repo["id"],
              :name => repo["display_name"],
              :description => repo["description"],
              :content_unit_counts => repo["content_unit_counts"],
              :type => REPO_TYPE_RPM,
              :last_unit_removed => repo["last_unit_removed"],
              :last_unit_added => repo["last_unit_added"],
              :distributor => distributor,
              :importer => importer,
            }
            #puts repos
          when REPO_TYPE_PUPPET
            puppet_distributor = repo["distributors"].select{ |d| d["distributor_type_id"] == 'puppet_distributor'}[0]
            distributor =nil
            if puppet_distributor
              distributor = {
                :auto_publish => puppet_distributor["auto_publish"],
                :last_publish => puppet_distributor["last_publish"],
                :config => puppet_distributor["config"]
              }
            end
            repo_data={
              :id => repo["id"],
              :name => repo["display_name"],
              :description => repo["description"],
              :content_unit_counts => repo["content_unit_counts"],
              :type => REPO_TYPE_PUPPET,
              :last_unit_removed => repo["last_unit_removed"],
              :last_unit_added => repo["last_unit_added"],
              :distributor => distributor
            }
          else
          end
          repo_data
      else
        raise "Exception: cannot get repository detail: response code :#{code}"
      end
    end#list_repo_details

    def create_rpm_repo(repo_id:, display_name: nil , description: '', feed_url: nil, relative_url: nil, serve_http: true, serve_https: false, auto_publish: false)

      if relative_url.nil? || relative_url.strip.length<1
        raise "Invalid relative_url"
      end
      importer = Runcible::Models::YumImporter.new
      importer.feed = feed_url if feed_url
      yum_distributor = Runcible::Models::YumDistributor.new relative_url, serve_http, serve_https
      yum_distributor.auto_publish = auto_publish
      yum_distributor.id = 'yum_distributor'

      export_distributor = Runcible::Models::ExportDistributor.new serve_http, serve_https, relative_url
      export_distributor.auto_publish = auto_publish
      export_distributor.id = 'export_distributor'

      optional = {
        :display_name => display_name,
        :description => description,
      }

      begin
        version = get_version #get server version
        # pulp issue https://pulp.plan.io/issues/1520
        if Gem::Version.new('2.8.0') > Gem::Version.new(version)
          def export_distributor.config
            to_ret = as_json
            to_ret.delete('auto_publish')
            to_ret.delete('id')
            to_ret.delete('relative_url')
            to_ret
          end
        end
        response = client.extensions.repository.create_with_importer_and_distributors(repo_id, importer, [yum_distributor, export_distributor], optional)
        code=response.code
        body=response.body
        puts "code #{code}"
        case code
        when 201
          return true
        default
          raise "Operation failed, response code:#{code}, #{response}"
        end
      rescue Exception => e
        raise "Failed to create repo, #{e.message}"
      end
    end

    def create_puppet_repo(repo_id:, display_name: nil , description: '', feed_url: nil, queries: nil,  remove_missing: false,  serve_http: true, serve_https: false, auto_publish: false)
      importer = Runcible::Models::PuppetImporter.new
      importer.feed = feed_url if feed_url
      importer.remove_missing = remove_missing
      importer.queries = queries
      puppet_distributor = Runcible::Models::PuppetDistributor.new nil, serve_http, serve_https
      puppet_distributor.auto_publish = auto_publish
      puppet_distributor.id = 'puppet_distributor'

      optional = {
        :display_name => display_name,
        :description => description,
      }

      begin
        response = client.extensions.repository.create_with_importer_and_distributors(repo_id, importer, [puppet_distributor], optional)
        code=response.code
        body=response.body
        puts "code #{code}"
        case code
        when 201
          return true
        default
          raise "Operation failed, response code:#{code}, #{response}"
        end
      rescue Exception => e
        raise "Failed to create repo, #{e.message}"
      end
    end

    def publish_repo!(forge_id)
      message = "Publish #{forge_id} submitted successfully"
      begin
        publish_response=client.extensions.repository.publish_all(forge_id)

        ##
        ## Runcilbe does not proper include respone code
        ##
        presponse=JSON.parse(publish_response.to_json)
        if presponse.nil? || presponse[0]["spawned_tasks"].length<1
          raise "Exception: repo publish requeste failed, response : #{publish_response}"
        end
        publish_response
      rescue StandardError => e
        raise "Excpetion: Failed to publish, #{e.message}"
      end
    end#publish_repo

    def sync_repo!(forge_id)
      begin
        sync_response=client.extensions.repository.sync(forge_id)
        ##
        ## Runcilbe does not proper include respone code
        ##
        if sync_response.code !=201 && !"#{sync_response.code}".start_with?("20")
          if sync_response.code == 404
            raise "repository #{forge_id} not found"
          end
          raise "Sync #{forge_id} failed with http response code: #{sync_response.code}"
        end
        sync_response
      rescue StandardError => e
        raise "Excpetion: Failed to Sync, #{e.message}"
      end
    end#sync_repo

    def sync_status(forge_id)
      begin
        sync_response=client.extensions.repository.sync_status(forge_id)
        ##
        ## Runcilbe does not proper include respone code
        ##
        if sync_response.code !=200 && !"#{sync_response.code}".start_with?("20")
          if sync_response.code == 404
            raise "repository #{forge_id} not found"
          end
          raise "Sync #{forge_id} failed with http response code: #{sync_response.code}"
        end
      rescue StandardError => e
        raise "Excpetion: Failed to Sync, #{e.message}"
      end
    end

    def delete_rpm_newer!(forge_id, name, version, release, arch, auto_publish=false)
      criteria = get_rpm_unit_ass_criteria(name, version, release, arch)
      begin
        unassociate_response=client.resources.repository.unassociate_units(forge_id, criteria)
        #pulp api documented response code
        if unassociate_response.code !=202
          raise "Exception: cannot unassociate unit, response code: #{unassociate_response.code}"
        end
        if auto_publish
          publish_response=client.extensions.repository.publish_all(forge_id)
          presponse=JSON.parse(publish_response.to_json)
          if presponse.nil? || presponse[0]["spawned_tasks"].length<1
            raise "Exception: repo publish requeste failed, response : #{publish_response}"
          end

          # if publish_response.code !=202 && !"#{publish_response.code}".start_with?("20")
          #    raise "Exception: repo publish requeste failed, response code : #{publish_response.code}"
          # end
        end
      rescue StandardError => e
        raise "Error delete rpm pakcage older than #{name}-#{version}-#{release} from repo #{forge_id}: #{e.message}"
      end
    end#delete_rpm_newer

    def copy_rpm_between_repo!(from_repo, to_repo, name, version, release, arch, delete_new=false, auto_publish=true)
      search_params=get_rpm_search_params(name, version, release, arch)
      begin
        unit_search_response=client.resources.unit.search('rpm', search_params[:criteria], search_params[:optional])

        if unit_search_response.code !=200
          raise "Exception: Cannot find unit"
        end
        found_units=JSON.parse(unit_search_response.to_json)
        unit_id=found_units.first['_id']

        copy_unit_ids= {
            :ids => [unit_id]
        }
        copy_response=client.extensions.rpm.copy(from_repo, to_repo, copy_unit_ids)
        if copy_response.code !=200 && copy_response.code !=202
          raise "Exception: unit copy failed with code #{copy_response.code}"
        end

        if(delete_new)
          delete_rpm_newer!(to_repo, name, version, release, arch,  auto_publish)
        end
      rescue StandardError => e
        raise "Exception: Error copy module between repo: #{e.to_s}"
      end
    end#copy_rpm_between_repo

    def delete_puppet_newer!(forge_id, author, name, version, auto_publish=false)
      #/pulp/api/v2/repositories/<repo_id>/actions/unassociate/
      criteria = get_puppet_unit_assoc_criteria(author, name, version)
      begin
        unassociate_response=client.resources.repository.unassociate_units(forge_id, criteria)
        #pulp api documented response code
        if unassociate_response.code !=202
          raise "Exception: cannot unassociate unit, response code: #{unassociate_response.code}"
        end
        if auto_publish
          publish_response=client.extensions.repository.publish_all(forge_id)

          presponse=JSON.parse(publish_response.to_json)
          if presponse.nil? || presponse[0]["spawned_tasks"].length<1
            raise "Exception: repo publish requeste failed, response : #{publish_response}"
          end
          # if publish_response.code !=202 && !"#{publish_response.code}".start_with?("20")
          #    raise "Exception: repo publish requeste failed, response code : #{publish_response.code}"
          # end
        end
      rescue StandardError => e
        raise "Exception: Error delete module newer than #{author}-#{name}-#{version} from repo #{forge_id}: #{e.message}"
      end
    end#function

    def copy_puppet_between_repo!(from_repo, to_repo, author,name, version, delete_new=false, auto_publish=true)
      search_params = get_puppet_search_params(author, name, version)
      begin
        puts "search_params :#{search_params}"
        unit_search_response=client.resources.unit.search("puppet_module", search_params[:criteria], search_params[:optional])
        if unit_search_response.code !=200
          raise "Exception: unit search faild with code #{unit_search_response.code}"
        end
        puts "search_response : #{unit_search_response.to_json}"
        found_units=JSON.parse(unit_search_response.to_json)
        unit_id=found_units.first['_id']

        unless unit_id && unit_id.length >0
          raise "No puppet module found to copy"
        end
        copy_unit_ids= {
            :ids => [unit_id]
        }
        copy_response=client.extensions.puppet_module.copy(from_repo, to_repo, copy_unit_ids)
        if copy_response.code !=200 && copy_response.code !=202
          raise "Exception: unit copy failed with code #{copy_response.code}"
        end
        if(delete_new)
          delete_puppet_newer!(to_repo,author, name, version, auto_publish )
        end
      rescue StandardError => e
        raise "Error copy module between repo: #{e.to_s}"
      end
    end# copy puppet

    def delete_repository(repo_id)
      begin
        response=client.resources.repository.delete(repo_id)
        case response.code
        when 202
          true
        when 404
          rase "Repository does not exist."
        default
          raise "Failed to delete repository."
        end
      rescue StandardError => e
        raise "Excpetion: Failed to delete, #{e.message}"
      end
    end

    ##### private functions #####
    private
    def get_rpm_search_params (name, version, release, arch)
      search_optional= {
          :include_repos => true
      }
      search_criteria ={
        :filters => {
            :name => name,
            :version=> version,
            :release => release,
            :arch => arch,
        },
        :sort => [["version", "descending"]],
        :limit => 1,
        :fields => ["name", "version", "release", "arch"]
      }

      search_params= {
        :optional => search_optional,
        :criteria => search_criteria,
      }

      return search_params
    end# get_rpm_search_params

    def get_rpm_unit_ass_criteria(name, version, release, arch)
       #/pulp/api/v2/repositories/<repo_id>/actions/unassociate/
      criteria= {
        :filters => {
            :unit => {
              '$and' => [
                  {:name => name},
                  {:arch => arch},
                  '$or' => [
                    :version=> {
                      '$gt' => version
                    },
                    :release => {
                      '$gt' => release
                    }
                  ]
              ]
            }
        },
        :type_ids => ["rpm"]
      }

      return criteria
    end#function

    def get_puppet_search_params(author, name, version)
      search_optional= {
          :include_repos => true
      }
      search_criteria ={
        :filters => {
            :author => author,
            :name => name,
            :version=> version
        },
        :sort => [["version", "descending"]],
        :limit => 1,
        :fields => ["author", "name", "version"]
      }

      search_params={
        :criteria =>  search_criteria,
        :optional => search_optional,
      }

      return search_params
    end#function
    def get_puppet_unit_assoc_criteria(author, name, version)
       criteria= {
        :filters => {
            :unit => {
              '$and' => [
                  {:name => name},
                  {:author => author},
                  :version=> {
                    '$gt' => version
                  }
              ]
            }
        },
        :type_ids => ["puppet_module"]
      }
      return criteria
    end#function


  end#module
end#module
